### PR TITLE
CLI: Add Jetpack Test Command

### LINKED ISSUE #20761: CLI: Suggestion: Add a new Jetpack Docker command to run PHPUnit only for a specific project
### What

Currently we have a `Jetpack docker phpunit` command that run PHPUnit tests only for Jetpack. Since now the monorepo includes several plugin, it could be good to have a similar command to run PHPUnit tests for a specific project

### How

Currently at least for the Boost plugin PHPunit can be ran in 2 ways:

- from outside Docker using `cd projects/plugins/boost; composer test-php`
- from inside Docker (kind of workaround) `jetpack docker exec -- sh -c "composer -d wp-content/plugins/boost test-php"`

Could it be beneficiary/useful to have a similar command to the `Jetpack docker phpunit` (which is only specific to Jetpack) but for a specific project instead. Maybe the command could be something like

- `jetpack docker plugins/boost phpunit`
 
This would be aligned with the other Jetpack CLI commands for specific projects such as for example:

- `jetpack install plugins/boost`
- `jetpack watch plugins/boost`

This requested suggestion can raise these 2 additional questions:
- What is the advantages to run the PHPunit for projects from Docker? Maybe consistency of environment for developers to run PHPunit tests?
- If we introduce a new command such as `jetpack docker PROJECT/PATH phpunit` what would become of the behaviour of the `Jetpack docker phpunit` command. Should it still carries on to run PHPunit tests only for Jetpack, or run PHPunit tests for ALL projects (maybe difficult to implement...)

Feel free to discard this request if really not relevant or not a priority since there are alternative solutions at the moment.

### PR DESCRIPTION
**Pull request by sdixon194**
Fixes #20761

#### Changes proposed in this Pull Request:

* Adds a `jetpack test` command that lets you run scripts defined in the project's composer.json files. 

Surely missing a few use cases here, but hopefully this is a good start to make running tests easier.

#### Jetpack product discussion

#20761

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Pull down this PR and run `jetpack test`
* Follow the prompts to run the test you want.
* Accepts two parameters:
	* `project` in the form of `project/name`, ie `plugins/jetpack`
	* `test,` accepting `js`, `php`, or `coverage`
* ie, `jetpack test plugins/jetpack js` should run JavaScript tests.

### CODE REVIEW

**Comment by kraftbj** [comment]
I generally like this. It would be good if the CLI (or something) to know if the test needs to run within Docker and run that. I'm not sure what's the best way to do that, e.g. `jetpack test plugins/jetpack php` fails.

**Comment by sdixon194** [comment]
> I generally like this. It would be good if the CLI (or something) to know if the test needs to run within Docker and run that. I'm not sure what's the best way to do that, e.g. jetpack test plugins/jetpack php fails.

Do we have an example of a project that requires Docker to run? I poked around but couldn't readily find one.

**Review by kraftbj — DISMISSED** [review]

File: tools/cli/commands/test.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * Runs the test script.
+ *
+ * @param {object} argv - the arguments passed.
+ */
+async function runTest( argv ) {
+	console.log( chalk.green( `Running ${ argv.testScript } tests for ${ argv.project }` ) );
+	child_process.spawnSync( 'composer', [ 'run', argv.testScript ], {
+		stdio: 'inherit',
+		cwd: __dirname + '/../../../' + `projects/${ argv.project }`,
```
Comment (anomiex) [inline tools/cli/commands/test.js:128]: Everywhere else we do like
```suggestion
		cwd: path.resolve( `projects/${ argv.project }` ),
```
That means it'll run the command in the correct place even if somehow it's running a copy of the CLI from elsewhere.

**Review by anomiex — COMMENTED** [review]
> Do we have an example of a project that requires Docker to run? I poked around but couldn't readily find one.

I think the PHP and coverage tests for Jetpack itself are the only ones that require it at the moment. But other plugins could too, in the CI test runner we set them all up with a WordPress install.

Probably the only way to reliably detect it will turn out to be setting `.extra.jetpack-cli.php-tests-in-docker` to true in the plugin's composer.json or something like that.

File: tools/cli/commands/test.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+ * @param {argv}  argv - the arguments passed.
+ * @returns {object} argv
+ */
+export async function promptForTest( argv ) {
+	const response = await inquirer.prompt( [
+		{
+			type: 'list',
+			name: 'test',
+			message: 'What test are you trying to run?',
+			choices: [ 'js', 'php', 'coverage' ],
```
[the remark this answers is not in the document]
Comment (sdixon194) [reply tools/cli/commands/test.js]: I like it! Done.

File: tools/cli/commands/test.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+			return 'test-php';
+		case 'coverage':
+			return 'test-coverage';
+		default:
+			// shouldn't ever happen, but being safe.
+			console.log(
+				chalk.red( `No ${ argv.test } script located in ${ argv.project }'s composer.json file!` )
+			);
+			process.exit();
+	}
```
Comment (anomiex) [inline tools/cli/commands/test.js:132]: This seems to have some odd failure cases. If someone adds a `test-e2e` to their composer.json, `promptForTest()` will offer "e2e" as an option only for this to claim it doesn't exist. And on the other hand, if someone supplies "php" on the command line for a non-PHP package it'll accept it here only to error out later.

Why not just use `getTests()` here too?
```suggestion
	const tests = await getTests( argv.project );
	if ( tests.includes( argv.tests ) ) {
		return 'test-' + argv.tests;
	} else {
		console.log(
			chalk.red( `No test-${ argv.test } script located in ${ argv.project }'s composer.json file!` )
		);
		process.exit( 1 );
	}
```

File: tools/cli/commands/test.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+				.positional( 'project', {
+					describe:
+						'Project in the form of type/name, e.g. plugins/jetpack, or type, e.g. plugins.',
+					type: 'string',
+				} )
+				.positional( 'test', {
+					alias: 't',
+					describe: 'The test to run',
+					type: 'Array',
+					choices: [ 'js', 'php', 'coverage' ],
```
Comment (anomiex) [inline tools/cli/commands/test.js:37]: As with the comment below, it'd be nice not to restrict the possible tests here if we're not doing so in `promptForTest()`.

File: tools/cli/commands/test.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+ * Gets list of tests available for chosen projects.
+ *
+ * @param {string} project - the project we want tests for..
+ * @returns {object} argv.
+ */
+async function getTests( project ) {
+	const composerJson = await readComposerJson( project );
+	if ( ! composerJson.scripts ) {
+		console.log( chalk.red( `No tests found in ${ project }'s composer.json file!` ) );
+		process.exit();
```
Comment (anomiex) [inline tools/cli/commands/test.js:100]: Should probably indicate a failure of some sort.
```suggestion
		process.exit( 1 );
```

File: tools/cli/commands/test.js [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	let tests = Object.keys( composerJson.scripts ).filter( test => test.startsWith( 'test-' ) );
+	tests = tests.map( test => test.substring( 5 ) );
+	if ( tests.length === 0 ) {
+		console.log( chalk.red( `No tests found in ${ project }'s composer.json file!` ) );
+		process.exit();
```
Comment (anomiex) [inline tools/cli/commands/test.js:107]:
```suggestion
		process.exit( 1 );
```

File: tools/cli/commands/test.js [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+ * Runs the test script.
+ *
+ * @param {object} argv - the arguments passed.
+ */
+async function runTest( argv ) {
+	console.log( chalk.green( `Running ${ argv.testScript } tests for ${ argv.project }` ) );
+	child_process.spawnSync( 'composer', [ 'run', argv.testScript ], {
+		stdio: 'inherit',
+		cwd: path.resolve( `projects/${ argv.project }` ),
+	} );
```
Comment (anomiex) [inline tools/cli/commands/test.js:145]: You should relay the exit code from the subprocess.
```suggestion
	const res = child_process.spawnSync( 'composer', [ 'run', argv.testScript ], {
		stdio: 'inherit',
		cwd: path.resolve( `projects/${ argv.project }` ),
	} );
	process.exitCode = res.status;
```

**Comment by sdixon194** [comment]
In line comments should be addressed 👍

**Review by anomiex — APPROVED** [review]
Looks good!

**Pull request merged**